web-server-doc/web-server/scribblings/dispatch-server.scrbl

```diff
@@ -54,55 +68,274 @@ ports (e.g., to implement SSL) as used by the dispatch server.
+  Passed to @racket[make-safety-limits].
+  @history[#:changed "1.6"
+           @elem{Loosened contract for consistency with @racket[make-safety-limits].}]
  }
- @defthing[dispatch (-> connection? any/c void)]{How to handle requests.}
 }
 
+
+@section{Safety Limits}
+@defmodule[web-server/safety-limits]
+
+@deftogether[
+ (@defproc[(safety-limits? [v any/c]) boolean?]
+   @defproc[(make-safety-limits
+             [#:max-waiting max-waiting exact-nonnegative-integer? 511]
+             [#:initial-connection-timeout initial-connection-timeout timeout/c 60]
+             [#:request-read-timeout request-read-timeout timeout/c 60]
+             [#:max-request-line-length max-request-line-length nonnegative-length/c
+              (code:line (* 8 1024) (code:comment #,(elem "8 KiB")))]
+             [#:max-request-headers max-request-headers nonnegative-length/c 100]
+             [#:max-request-header-length max-request-header-length nonnegative-length/c
+              (code:line (* 8 1024) (code:comment #,(elem "8 KiB")))]
+             [#:max-request-body-length max-request-body-length nonnegative-length/c
+              (code:line (* 1 1024 1024) (code:comment #,(elem "1 MiB")))]
+             [#:max-request-files max-request-files nonnegative-length/c 100]
+             [#:max-request-file-length max-request-file-length nonnegative-length/c
+              (code:line (* 10 1024 1024) (code:comment #,(elem "10 MiB")))]
+             [#:request-file-memory-threshold request-file-memory-threshold nonnegative-length/c
+              (code:line (* 1 1024 1024) (code:comment #,(elem "1 MiB")))]
+             [#:response-timeout response-timeout timeout/c 60]
+             [#:response-send-timeout response-send-timeout timeout/c 60])
+            safety-limits?]
+   @defthing[nonnegative-length/c flat-contract?
+             #:value (or/c exact-nonnegative-integer? +inf.0)]
+   @defthing[timeout/c flat-contract?
+             #:value (>=/c 0)])]{
+ The web server uses opaque @deftech{safety limits} values, recognized
+ by the predicate @racket[safety-limits?], to encapsulate
+ policies for protection against misbehaving or malicious clients and servlets.
+ Construct @tech{safety limits} values using @racket[make-safety-limits],
+ which supplies reasonably safe default policies that should work for most applications.
+ See the @elemref["safety-limits-porting"]{compatability note} and
+  @racket[make-unlimited-safety-limits] for further details.
+
+ The arguments to @racket[make-safety-limits] are used as follows:
+ @itemlist[
+ @item{The @racket[max-waiting] argument is passed to @racket[tcp-listen]
+   to specify the maximum number of client connections that can be waiting for acceptance.
+   When @racket[max-waiting] clients are waiting for acceptance, no new client connections can be made.
+   }
+  @item{The @racket[initial-connection-timeout] specifies the initial timeout,
+   in seconds, given to each connection.
+   }
```

Comment:
@Bogdanp I think it would be useful to say more about how `initial-connection-timeout` interacts with the other timeouts, but I'm not 100% sure I've worked it all out myself.

Reply:
I think the "timeout plan" at the top of `timer.rkt` is a fairly accurate description, although it should really mention which timeouts the timer gets reset to at every step. I think the `initial-connection-timeout` is completely superseded by the more fine-grained timeouts and is now only really used as the maximum time before the web server must start processing a request. Unless it's set to 0 or the server is under extreme load, it should never be reached.

Reply:
Thanks for pointing me to this! From looking at the old behavior: https://github.com/racket/web-server/blob/ba300fde5f88a2d0883d37eaaba9ee140d425896/web-server-lib/web-server/private/timer.rkt#L107-L110

I think it would make more sense to remove `#:initial-connection-timeout` from `safety-limits`, pass deprecated `#:initial-connection-timeout` arguments as `#:request-read-timeout`, and set the initial timeout of the connection to the `#:request-read-timeout`. Does that seem right to you?

Reply:
Yes, I think that sounds reasonable.